packages/tpetra/core/src/Tpetra_withLocalAccess.hpp

```diff
@@ -315,25 +585,67 @@ namespace Tpetra {
       /// on allocating temporary space, copying from device to host,
       /// etc.  This implies that implementations must be able to
       /// construct "null" / empty master local objects.
-      LocalAccess<GlobalObjectType, MemorySpace, am>
+      this_type
       valid (const bool thisIsValid) const {
-        return {this->G_, this->space_, thisIsValid};
+        return {G_, getExecutionSpace(), getMemorySpace(),
+                thisIsValid};
       }
 
-      /// \brief Declare intent to access this object's local data in
+      /// \brief Declare intent to access this object's local data on
       ///   a specific (Kokkos) memory space (instance).
       template<class NewMemorySpace>
-      LocalAccess<GlobalObjectType, NewMemorySpace, am>
-      on (NewMemorySpace space) const {
-        return {this->G_, space, this->valid_};
+      typename std::conditional<
+        is_execution_space_specified,
```

Comment:
If the user didn't specify an execution space at which to view the local data, we don't "pre-constrain" it.